### ISSUE TITLE
FW4 issues (official support for OpenWRT 22.03.2/Firewall4?)

### ISSUE THREAD

**Issue by JoeDat** [openNDS/openNDS#412]
Writing to understand if full support for OpenWRT 22.03.x, and more precisely, firewall4 is expected.  Based on information I've found in the OpenWRT github repos as well as this one, I'm lead to believe that it should work (with an understanding that "functionality" and "support" are vastly different things).  However, the opennds docs state testing was completed up through 21.2.x, with some statements suggesting that support for 21.2.x+ is implied.  If 22.03.x is not officially supported yet on the latest opennds release, I can certainly live with waiting until it is and we can close this out.

If it is supported, then I'm running into an issue after upgrading from 21.2.x to 22.03.3 and looking for some help.  Prior to the upgrade, opennds worked flawlessly.  Now, it doesn't appear as though the fw rules implemented by the service are functional at startup.  The redirects in place for ports 80 and 443 do not send the browser to the CPD page.  I also cannot access the CPD page at port 2050 unless I create an explicit accept rule for the src zone and destination port.  Also, the rules are shown as legacy iptables rules as opposed to nftables in the OpenWRT luci, and I'm unsure if this is intended. 

I tried removing all opennds customization and just went with the default opkg config, but no dice.  Happy to supply whatever additional info is needed for troubleshooting.  

Running OpenWrt 22.03.2 on a WRT3200ACM, with OpenNDS 9.9.1.  I am using a custom themespec script, but I'm seeing similar issues with a default package config.

**Comment by bluewavenet** [comment]
@JoeDat 

> full support for OpenWRT 22.03.x, and more precisely, firewall4

To be clear, openNDS 9.9.0 and onwards fully supports openWrt 22.03.x and is incompatible with earlier versions of the operating system.
openNDS does not use FW4 but works seamlessly with it in the same way as older, pre-9.9.x, was compatible with FW3.

_Thanks for pointing out that the documentation has not been updated to reflect this! It will be updated in due course._

>  Also, the rules are shown as legacy iptables rules as opposed to nftables in the OpenWRT luci

There is a known bug in the "iptables" package in 22.03, and Master.
If a package declares a dependency in iptables, it will result in iptables-zz-legacy being installed, usually resulting in the symptoms you are seeing.

openNDS specifically requires iptables-nft as an install dependency to work around the iptables install bug.
This workaround is fine unless some other package with just "iptables" as a dependency is also installed.
It is then possible, even likely, that **BOTH** iptables-nft and iptables-zzz-legacy will be installed together.

The result of this is is indeterminate, and quite often results in the luci warning message.

Check what is installed using:
`opkg list-installed | grep "iptables"`

**Comment by JoeDat** [comment]
Thanks @bluewavenet !   Sure enough, both iptables-zz-legacy and iptables-nft were both installed.  I've removed iptables-zz-legacy.  However, opennds is still implementing legacy rules.  I'm not sure how that's possible.  Package is definitely gone:

```
root@openwrt:/usr/share# opkg list-installed | grep "iptables"
iptables-mod-ipopt - 1.8.7-7
iptables-nft - 1.8.7-7
```
Log output on startup appears normal.  At least, the exit codes suggest successful runs of the iptables commands.  Ex:

```
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Executing command: iptables -V
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Reading command output
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: command output: [iptables v1.8.7 (nf_tables) ]
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Iptables mark Preauthenticated: 0x0
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Iptables mark Authenticated: 0x30000
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Iptables mark Trusted: 0x20000
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Iptables mark Blocked: 0x10000
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Executing command: iptables --wait -t mangle -I PREROUTING 1 -j MARK --or-mark 0x10000 > /dev/null 2>&1
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: iptables command [ -t mangle -I PREROUTING 1 -j MARK --or-mark 0x10000 ], return code [ 0 ]
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Executing command: iptables --wait -t mangle -D PREROUTING 1 > /dev/null 2>&1
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: iptables command [ -t mangle -D PREROUTING 1 ], return code [ 0 ]
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Kernel supports --or-mark.
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Executing command: iptables --wait -t filter -I FORWARD 1 -m mark --mark 0x10000/0x30000 -j REJECT > /dev/null 2>&1
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: iptables command [ -t filter -I FORWARD 1 -m mark --mark 0x10000/0x30000 -j REJECT ], return code [ 0 ]
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Executing command: iptables --wait -t filter -D FORWARD 1 > /dev/null 2>&1
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: iptables command [ -t filter -D FORWARD 1 ], return code [ 0 ]
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Kernel supports mark masking.
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Iptables mark op "--or-mark" and mark mask "/0x30000".
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: Executing command: iptables --wait -t mangle -N ndsTRU
Mon Nov 28 08:21:27 2022 daemon.debug opennds[18851]: iptables command [ -t mangle -N ndsTRU ], return code [ 0 ]
```

Example of what I see in luci.  I assume this is not expected:

![image](https://user-images.githubusercontent.com/8648894/204289733-e3dfda1f-d597-489b-bcc0-8793e2e0eb23.png)

**Comment by bluewavenet** [comment]
@JoeDat 
Yes you normally need to reboot to clear out all the legacy mess....

What other packages did you install? One or more of them resulted in the iptables-zz-legacy package being installed.
The workaround is to install iptables-nft **_before_** anything else.

**Comment by JoeDat** [comment]
@bluewavenet  Just reporting back that I'm still getting legacy rules from opennds on a fresh install of 22.03.3 without any custom configuration.  iptables isn't provided in the base openwrt package.  I explicitly installed iptables-nft first, then opennds.  These are the only two cusotm packages I've added on to the base build. 

It could be something in the base build of the mvebu/cortexa9 target platform, but nothing is standing out.  I'm not convinced of this since iptables isn't even available.  So, I'm not sure if the issue is in opennds or openwrt at this point.  I'm going to try 22.03.2 later on.  Perhaps something new introduced in this minor rev.

**Comment by bluewavenet** [comment]
@JoeDat 

> Example of what I see in luci. I assume this is not expected:

I don't think the screenshot you show is an error. It is just showing that ndsTRU is empty and this is normal. It will contain details of any trusted clients you might add to the config that do not have to authenticate.

>  I explicitly installed iptables-nft first

The openNDS package automatically installs iptables-nfs. So you only need to do this explicitly.

If SQM is installed AFTER openNDS, it will not try to install "iptables".
The known bug is that if you do for example:
`opkg install iptables`
you will get `iptables-zz-legacy` instead of the required `iptables-nft`.

So if SQM is installed first, it calls for "iptables" as a dependency, and gets iptables-zz-legacy.
Then openNDS comes along and calls for iptables-nft, resulting in both being installed.

**Comment by JoeDat** [comment]
I appreciate you working with me on this.  I suppose the question I need to confirm is if I should expect the FW to report legacy rules only when starting OpenNDS.  Right now, I have a vanilla install of OpenWRT 22.03.2 installed, only with OpenNDS installed using a default config.  I'm not being redirected to the CPD login, nor is the landing page accessible.  As mentioned previously, I can confirm it's something FW related by manually adding the forwarding and traffic rules to the FW (via luci)

Before starting OpenNDS, luci isn't showing any warnings regarding legacy rules.  After I start OpenNDS, the warning appears.  Conversely, if I stop OpenNDS, the warning is removed.  With only having iptables-nft installed, is this expected?  If it is, I've got something else going on.

**Comment by bluewavenet** [comment]
@JoeDat 
`I'm not being redirected to the CPD login, nor is the landing page accessible. `

This shows you have something wrong as it is working elsewhere.
Please show the output of `uci show opennds`

**Comment by JoeDat** [comment]
This isn't my normal opennds config.  I've since restored my openwrt backup which re-implements my custom vlan setup.  I don't have a br-lan.  The usual network this was configured for is br-lan.3 which is a guest wlan.  However,  I am keeping opennds as vanilla as possible to rule out any other customization.  Needed to set br-lan.3 for the service to startup.  

```
root@openwrt:/etc/config# uci show opennds
opennds.@opennds[0]=opennds
opennds.@opennds[0].enabled='1'
opennds.@opennds[0].fwhook_enabled='1'
opennds.@opennds[0].maxclients='250'
opennds.@opennds[0].users_to_router='allow tcp port 53' 'allow udp port 53' 'allow udp port 67' 'allow tcp port 22' 'allow tcp port 80' 'allow tcp port 443'
opennds.@opennds[0].gatewayinterface='br-lan.3'
```

**Comment by bluewavenet** [comment]
@JoeDat 
There are two types of legacy rules and luci does not differentiate between them when it displays the warning message.

Type 1: Legacy rules created by iptables-legacy.
These rules are compatible with fw3 ie are raw "iptables". They are not translated into nftables rules and are not displayed in the fw4 nftables output. These "raw" legacy rules will still be active but may well clash with and cause problems when nftables rules are also in place. In this case the luci error message is valid. It is inadvisable to depend upon any such rules. 

Type 2: Legacy rules created by iptables-nft.
These rules are automatically translated into nftables rules and can be seen in the fw4 nftables output.
The legacy iptables rules are **_not_** active in raw format, instead their nftables translated versions are used instead. These translated rules are fully integrated with any other nftables rules and therefore cannot cause any og the problems that might arise with Type 1 above. The luci warning is in this case invalid.

Example Luci screenshot of nftables output for openNDS using iptables-nft (one authenticated client):

![Screenshot from 2022-11-29 08-28-37](https://user-images.githubusercontent.com/13632068/204478507-b5de7088-04d2-4c44-ab4e-63fe39cf2934.png)

**Comment by bluewavenet** [comment]
@JoeDat 
For looking at this issue, I have a freshly reflashed gl-inet mt300n-v2 with 22.03.2.
I have installed openNDS with opkg, and nothing else.
The luci screenshot above was taken on this device.

On connecting an Android phone, I immediately get the CPD splash page ie the default click to continue Themespec.
On clicking "Accept Terms of Service" I get the default "Thankyou page" with a "Continue" button.
On clicking "Continue" I get the authenticated user landing page with a link to the client account status page.
ie All is working as it should.

The output of `uci show opennds` is as follows:
```
opennds.@opennds[0]=opennds
opennds.@opennds[0].enabled='1'
opennds.@opennds[0].fwhook_enabled='1'
opennds.@opennds[0].maxclients='250'
opennds.@opennds[0].users_to_router='allow tcp port 53' 'allow udp port 53' 'allow udp port 67' 'allow tcp port 22' 'allow tcp port 80' 'allow tcp port 443'

```

The only difference I can see over yours is that you are attempting to bind to br-lan.3

Is br-lan.3 configured correctly?
Show the output of:
`ip addr`

Also, do:
`service opennds restart`

Wait about 10 seconds then show the output of:
`logread -e opennds | tail -n 30`

**Comment by JoeDat** [comment]
@bluewavenet 
So firstly, thanks for explaining the warning behavior in luci when it comes to legacy/nft and legacy/ipt tables.  I know it's not your product, so I appreciate you taking the time to do that.  It all makes much more sense to me now.

As for this issue, you lead me to the cause.  br-lan.3/guest default forwarding was configured to reject INPUT.   I switched the default forwarding for INPUT to accept, and automagically, it all started working.  I restored my opennds config and everything is back to normal. 

![image](https://user-images.githubusercontent.com/8648894/204532422-c74eb457-0bf3-40e2-8a69-dab80434984a.png)

This is definitely a deviation from my config on 21.02.3.  I went into my backup prior to the upgrade to make sure I'm not crazy and confirmed I had the default INPUT policy set to reject.   I can only assume this behavior has changed as a result of fw4?   Anyway, I'm not complaining.

Thanks so much again for your fantastic support of this product!  Ill keep this open unless you think there's a bug or room for improvement here in the way opennds should handle the default forwarding policies for the managed interface.  Otherwise, I'm all set!

**Comment by bluewavenet** [comment]
@JoeDat 

> br-lan.3/guest default policy was configured to reject INPUT

With the release of openNDS v9.3.0 in August 2021, the default openNDS rules were changed from "Allow" to "Passthrough". 
Prior to v9.3.0, openNDS defaulted to "Allow", so did not chain to the fw3 rules.

See:
https://opennds.readthedocs.io/en/stable/config.html#grant-access-for-authenticated-users-allow-and-passthrough

What version of openNDS were you previously using?

**Comment by JoeDat** [comment]
I had upgraded to openNDS v9.8.0 on 10/30/2022.  However, it's very possible that I only had preauthenticated clients on the network between that date and when I upgraded (and subsequently tested opennds again).  I'd have to go back through my logs to confirm or deny that.  However, I'm fairly certain that I tested the CPD after the upgrade, as I had noticed the css changes made from m prior installed version, which was likely significantly older.
